Modules/readline.c

```diff
@@ -1026,7 +1026,19 @@ on_startup_hook(void)
 {
     int r;
     PyGILState_STATE gilstate = PyGILState_Ensure();
-    r = on_hook(readlinestate_global->startup_hook);
+    PyObject* mod = PyState_FindModule(&readlinemodule);
+    if (mod == NULL) {
+        PyGILState_Release(gilstate);
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ImportError, "readline import initialization failure during startup hook");
```

Comment:
How about also unregistering the callback:

```suggestion
            PyErr_SetString(PyExc_ImportError, "readline import initialization failure during startup hook");
            rl_startup_hook = NULL;
```

or better:

```suggestion
            PyErr_SetString(PyExc_ImportError, "readline import initialization failure during startup hook");
            cleanup_readline();
```

Where `cleanup_readline()` is a new static function that undoes everything `setup_readline()` did, especially unregistering all the hooks.

Reply:
I went with the second option